File: backend/app/services/election_service.py
from typing import List, Optional
from app.repositories.election_repository import ElectionRepository
from app.models.election import ElectionResponse

class ElectionService:
    """Service layer for election-related business logic."""

    def __init__(self, repository: ElectionRepository = None):
        self.repository = repository or ElectionRepository()

    def list_elections(self) -> List[ElectionResponse]:
        """
        Fetches all elections and maps them to ElectionResponse models.
        """
        raw_elections = self.repository.get_all_elections()
        
        # Sort elections by year descending
        raw_elections.sort(key=lambda x: x.get("year", 0), reverse=True)

        return [
            ElectionResponse(
                id=item.get("PK", ""),
                year=item.get("year", 0),
                type=item.get("type", ""),
                category=item.get("category", "")
            )
            for item in raw_elections
        ]

    def get_election(self, election_id: str) -> Optional[ElectionResponse]:
        """
        Fetches a single election and maps it to an ElectionResponse model.
        """
        item = self.repository.get_election_by_id(election_id)
        if not item:
            return None

        return ElectionResponse(
            id=item.get("PK", ""),
            year=item.get("year", 0),
            type=item.get("type", ""),
            category=item.get("category", "")
        )

File: frontend/lib/services/election.service.ts
import { MLARepository } from "../repositories/mla.repository";
import { ElectionResponse } from "@/types/models";
import { DynamoDBWrapper } from "../dynamodb";
import { unstable_cache } from "next/cache";

export class ElectionRepository {
  private client: DynamoDBWrapper;

  constructor(tableName: string = process.env.ELECTIONS_TABLE || "knowyourmla_elections") {
    this.client = new DynamoDBWrapper(tableName);
  }

  async getAllElections() {
    return this.client.scan();
  }

  async getElectionById(electionId: string) {
    return this.client.get({
      Key: { PK: electionId, SK: "METADATA" },
    });
  }
}

export class ElectionService {
  private repository: ElectionRepository;

  constructor(repository?: ElectionRepository) {
    this.repository = repository || new ElectionRepository();
  }

  async listElections(): Promise<ElectionResponse[]> {
    return unstable_cache(
      async (): Promise<ElectionResponse[]> => {
        const rawElections = await this.repository.getAllElections();
        rawElections.sort((a, b) => parseInt(b.year || "0") - parseInt(a.year || "0"));

        return rawElections.map((item: any) => ({
          id: item.PK,
          year: parseInt(item.year || "0"),
          type: item.type,
          category: item.category,
        }));
      },
      ["all-elections-list"],
      { revalidate: 86400, tags: ["elections"] }
    )();
  }

  async getElection(electionId: string): Promise<ElectionResponse | null> {
    return unstable_cache(
      async (eId: string): Promise<ElectionResponse | null> => {
        const item = await this.repository.getElectionById(eId);
        if (!item) return null;

        return {
          id: item.PK,
          year: parseInt(item.year || "0"),
          type: item.type,
          category: item.category,
        };
      },
      ["election-by-id", electionId],
      { revalidate: 86400, tags: [`election-${electionId}`] }
    )(electionId);
  }
}
